### ISSUE TITLE
EBADF crash (uv__io_stop: Assertion `loop->watchers[w->fd] == w' failed)

### ISSUE THREAD

**Issue by dragonbane0** [brianc/node-postgres#1830]
**Version of Node**
Tested 8, 9 and 10.15.1 LTS Branch

**Version of Postgres**
PostgreSQL 10.5 (Ubuntu 10.5-0ubuntu0.18.04) on x86_64-pc-linux-gnu, compiled by gcc (Ubuntu 7.3.0-16ubuntu3) 7.3.0, 64-bit

**Version of node-postgres**
7.8.0

**Error**
```
node /index.js: ../deps/uv/src/unix/core.c:896: uv__io_stop: Assertion `loop->watchers[w->fd] == w' failed.
Unhandled exception:  { Error: write EBADF
    at WriteWrap.afterWrite [as oncomplete] (net.js:841:14)
  errno: 'EBADF',
  code: 'EBADF',
  syscall: 'write',
  client: 
   Client {
     _events: { error: [Function] },
     _eventsCount: 1,
     _maxListeners: undefined,
     connectionParameters: 
      ConnectionParameters {
        user: xxx,
        database: xxx,
        port: xxx,
        host: xxx,
        binary: false,
        ssl: false,
        client_encoding: '',
        replication: undefined,
        isDomainSocket: false,
        application_name: undefined,
        fallback_application_name: undefined,
        statement_timeout: false },
     user: xxx,
     database: xxx,
     port: xxx,
     host: xxx,
     replication: undefined,
     _Promise: [Function: Promise],
     _types: TypeOverrides { _types: [Object], text: {}, binary: {} },
     _ending: true,
     _connecting: false,
     _connected: true,
     _connectionError: false,
     _queryable: false,
     connection: 
      Connection {
        _events: [Object],
        _eventsCount: 19,
        _maxListeners: undefined,
        stream: [Socket],
        _keepAlive: false,
        lastBuffer: false,
        lastOffset: 0,
        buffer: null,
        offset: 1,
        encoding: 'utf8',
        parsedStatements: {},
        writer: [Object],
        ssl: false,
        _ending: true,
        _mode: 0,
        _emitMessage: false,
        _reader: [Object] },
     queryQueue: [],
     binary: false,
     processID: 4251,
     secretKey: xxxx,
     ssl: false,
     release: [Function: bound release],
     activeQuery: null,
     readyForQuery: false,
     hasExecuted: true } }
```

The assert is followed immediately by the EBADF unhandled exception error forcing us to auto-restart our production server. Sometimes just the "node /index.js: ../deps/uv/src/unix/core.c:896: uv__io_stop: Assertion `loop->watchers[w->fd] == w' failed" assert failure is thrown. Going off the presence of  ConnectionParameters referencing our DB connection string I have to assume this is related to node-postgres as we don't use anything else to interact with our DB. If there is a way to trace the source more closely I will gladly do so (since maybe Node itself is at fault)

This seems to be a race condition as it only started happening once we received substantial more traffic than usual. It only happens under load (several queries every second) and in various pieces of our program as far as I can tell with just log messages with different types of queries. Hard to track down since you need traffic to replicate it, but it happens roughly 3 times a day.

We tried Node 10 as it incorporates some fixes to libuv that ensure that instead of an assert you get a JS error, but this did not help. We still get the assert.

We also tried increasing Ubuntu's limit of max allowed file descriptors/opened files, which didn't help either. Memory/RAM is plenty available.

**Additional context:** 
We initialize node-postgres as follows in our code:

```
const pg = require('pg');

const sqlPool = new pg.Pool({
    user: xxx,
    host: xxx,
    database: xxx',
    password: xxx,
    port: xxx,
    idleTimeoutMillis: 0,
    max: 100
});
```

The entire app uses this single pool. We exclusively use sqlPool.query to interact with our DB. The DB has max_connections set to 200.

**Comment by Codolophier** [comment]
I know this is a year old thread, but did you ever find a solution?  I am seeing the same problem on:

- **Node**:  13.9.0
- **OS**: CentOS 7.7
- **PostgreSQL**: 10.12
- **Node Postgres**:  7.18.2

I've been through many of the same steps that you've specified such as increasing ulimits, memory, connection limits, etc., all to no avail.  On Windows, everything works "_perfectly_",  but on CentOS this crops up seemingly under the same manner of conditions that you've specified (e.g., lots of back-to-back select queries, inserts, updates, etc.).  Following is further information on the stack trace that I'm seeing (via ```process.on("uncaughtException",...)``` ):

```
Error: write EBADF
   at afterWriteDispatched (internal/stream_base_commons.js:154:25)
   at writeGeneric (internal/stream_base_commons.js:145:3)
   at Socket._writeGeneric (net.js:784:11)
   at Socket._write (net.js:796:8)
   at doWrite (_stream_writable.js:454:12)
   at writeOrBuffer (_stream_writable.js:436:5)
   at Socket.Writable.write (_stream_writable.js:327:11)
   at Connection._send (/home/prj/motor-wheel/node_modules/pg/lib/connection.js:228:24)
   at Connection.flush (/home/prj/motor-wheel/node_modules/pg/lib/connection.js:331:8)
   at Query._getRows (/home/prj/motor-wheel/node_modules/pg/lib/query.js:182:14) {
     errno: -9,
     code: 'EBADF',
     syscall: 'write'
   }
```

Any information or guidance that you - or anyone - can provide will be greatly appreciated.

**Comment by dragonbane0** [comment]
In my case most of the occurrences were actually caused by a npm module called "tmp-promise" which I used to get a random file name. I tried core dumps and stacktrackes, but apparently that module messed up something so deep causing race conditions in the node C++ core itself I was never able to pinpoint it down. After removing it the issues mostly stopped. 

So I would recommend evaluating all your dependencies, even if they seem innocent and unrelated.
